MapsetVerifier.Checks/Utils/TaikoUtils.cs

```diff
@@ -0,0 +1,199 @@
+        /// <list type="bullet">
+        /// <item><description>If the BPM is less than 110 BPM, the snap divisor is doubled.</description></item>
+        /// <item><description>If the BPM is greater than 270 BPM, the snap divisor is halved.</description></item>
+        /// <item><description>If the BPM is greater than 130 BPM, the snap divisor is divided by 1.5.</description></item>
+        /// </list>
+        /// </summary>
+        public static double GetNormalizedMsPerBeat(this UninheritedLine line)
+        {
+            double result = line.msPerBeat;
+
+            while (result <= 60000.0 / 270.0) // 270 BPM
+                result *= 2;
+
+            while (result >= (60000.0 / 110.0)) // 110 BPM
+                result /= 2;
+
+            while (result >= (60000.0 / 130.0)) // 130 BPM
+                result /= 1.5;
+
+            return result;
+        }
+        
+        /// <summary>
+        ///     Returns a list of timing lines where kiai toggles occur (kiai state changes).
+        /// </summary>
+        public static List<TimingLine> FindKiaiToggles(this List<TimingLine> timingLines)
+        {
+            List<TimingLine> kiaiToggles = new List<TimingLine>();
+
+            TimingLine? previousTimingLine = null;
+
+            foreach (TimingLine line in timingLines)
+            {
+                if ((previousTimingLine == null && line.Kiai) || (previousTimingLine != null && previousTimingLine.Kiai != line.Kiai))
+                {
+                    kiaiToggles.Add(line);
+                }
+                previousTimingLine = line;
+            }
+            return kiaiToggles;
+        }
+
+        /// <summary>
+        ///     Returns a list of timing lines where kiai toggles occur (kiai state changes) for the given beatmap.
+        /// </summary>
+        public static List<TimingLine> FindKiaiToggles(this Beatmap beatmap) => beatmap.TimingLines.FindKiaiToggles();
+        
+        /// <summary>
+        ///     Returns a list of timing lines where SV (slider velocity) changes occur.
+        /// </summary>
+        public static List<TimingLine> FindSvChanges(this List<TimingLine> timingLines)
+        {
+            List<TimingLine> svChanges = new List<TimingLine>();
+            
+            for (int i=0; i<timingLines.Count-1; i++)
+            {
+                TimingLine firstLine = timingLines[i];
+                TimingLine secondLine = timingLines[i+1];
+
+                if (firstLine.SvMult != secondLine.SvMult && firstLine.Offset != secondLine.Offset)
+                {
+                    svChanges.Add(secondLine);
+                }
+            }
+
+            return svChanges;
+        }
+
+        /// <summary>
+        ///     Returns a list of timing lines where SV (slider velocity) changes occur for the given beatmap.
+        /// </summary>
+        public static List<TimingLine> FindSvChanges(this Beatmap beatmap) => beatmap.TimingLines.FindSvChanges();
```

Comment:
Could make this part of the Beatmap class where we have a function that returns all SV changes as this is probably something Mania also needs?

Reply:
Hmmm. I think something like this is non-essential to basic parsing, so I think I'd still prefer it separate (I guess I don't feel super strongly, but that's just where I'm leaning). Since it's implemented as an extension method, it can be called in a similar way to if it were implemented within the Beatmap class already.

But moving it to a common place that isn't labelled specific to taiko seems good to me.